### PR TITLE
refactor(abigen): source

### PR DESCRIPTION
**Pull request by DaniPopes**
## Motivation

Parse test was not working

Crate was always built with reqwest, tokio even when most of the time it is not used. Ex: build script depending on `ethers-contract-abigen` that only wants to build local files now also gets all of reqwest, and tokio

TODO: upstream this change also in `ethers-contract` -- would be nice if we had #1963.

## Solution

Split into normal and online modules

Reqwest depends on Tokio so there's no extra dependencies

There should be no breaking changes + tests work now

## PR Checklist

-   [x] Added Tests
-   [x] Added Documentation
-   [ ] Updated the changelog
-   [ ] Breaking changes

### CODE REVIEW

File: ethers-contract/ethers-contract-abigen/Cargo.toml [hunk 1]
```diff
@@ -15,23 +15,30 @@ keywords = ["ethereum", "web3", "celo", "ethers"]
 
 [dependencies]
 ethers-core = { version = "^1.0.0", path = "../../ethers-core", features = ["macros"] }
+ethers-etherscan = { path = "../../ethers-etherscan", default-features = false, optional = true }
```
Comment (mattsse) [inline ethers-contract/ethers-contract-abigen/Cargo.toml:18]: why do we need etherscan here?

**Review by mattsse — CHANGES_REQUESTED** [review]
I'm a bit confused.

does this PR fix a feature issue and also add support for etherscan support to the contract crate?

if so, please split them into separate PRs, otherwise, this is very confusing.

> Parse test was not working

which test?

> Crate was always built with reqwest, tokio even when most of the time it is not used.

which crate? if ethers-contract, this makes sense since reqwest was default?

File: ethers-contract/ethers-contract-abigen/Cargo.toml (hunk 1 above)
Comment (DaniPopes) [reply ethers-contract/ethers-contract-abigen/Cargo.toml:18]: we have duplicate code for parsing the get request when we could just use the specialized crate in the same project

**Comment by DaniPopes** [comment]
> which test?

There's 1 test for `Source`, `parse_sources`, and it didn't work

> does this PR fix a feature issue and also add support for etherscan support to the contract crate?

> which crate? if ethers-contract, this makes sense since reqwest was default?

`ethers-contract-abigen` fails/panics at runtime in many cases if reqwest is not enabled or is wasm. In my opinion this whole thing is currently not handled well and that is what this PR aims to do. 
I also think that `online` features (with tons of extra dependencies) should not be enabled by default, with offline being an opt in, but the other way around. But this can be another issue/pr

**Comment by mattsse** [comment]
I see, still let's do this in two steps?

**Comment by DaniPopes** [comment]
Not sure how to split this? Since it's not really adding support for Etherscan, it's just moving to using ethers-etherscan instead + refactoring the whole thing in the process

**Review by gakonst — APPROVED** [review]
good with me. the ethers_etherscan import is OK not against it

**Pull request merged**